File: python/992_SubarrayswithKDifferentIntegers.py
from unittest import TestCase
# https://leetcode.com/problems/subarrays-with-k-different-integers/
import collections


class MultiSet:
    def __init__(self):
        self.size = 0
        self.cnt = collections.defaultdict(lambda : 0)

    def add(self, v):
        self.cnt[v] += 1
        if self.cnt[v] == 1:
            self.size += 1

    def remove(self, v):
        self.cnt[v] -=1
        if self.cnt[v] == 0:
            self.size -= 1


class SubarrayswithKDifferentIntegers(TestCase):
    def subarraysWithKDistinct(self, A: 'List[int]', K: 'int') -> 'int':
        sj, sk = MultiSet(), MultiSet()
        i, j, k, n, res = 0, 0, 0, len(A), 0
        while i < n:
            while sj.size < K and j < n:
                sj.add(A[j])
                j += 1
            if sj.size < K:
                break
            while sk.size <= K and k < n:
                sk.add(A[k])
                k += 1
            res += k - j + (1 if sk.size == K else 0)
            sj.remove(A[i])
            sk.remove(A[i])
            i += 1
        return res

    def test1(self):
        self.assertEqual(7, self.subarraysWithKDistinct([1,2,1,2,3], 2))

    def test2(self):
        self.assertEqual(3, self.subarraysWithKDistinct([1,2,1,3,4], 3))

File: kotlin/src/main/kotlin/992.SubarrayswithKDifferentIntegers.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/subarrays-with-k-different-integers/


class SubarrayswithKDifferentIntegers{
  data class MultiSet(var size: Int = 0){
    val cnt = hashMapOf<Int, Int>()
    fun add(v: Int){
      cnt[v] = cnt.getOrDefault(v, 0) + 1
      if(cnt[v] == 1)
        size ++
    }
    fun remove(v: Int){
      cnt[v] = cnt[v]!! - 1
      if(cnt[v] == 0)
        size --
    }
  }

  fun subarraysWithKDistinct(A: IntArray, K: Int): Int {
    var (j, k) = 0 to 0
    val (sj, sk) = MultiSet() to MultiSet()
    var res = 0
    for((i, v) in A.withIndex()){
      sj.add(v)
      sk.add(v)
      while(sj.size > K)
        sj.remove(A[j++])
      while(sk.size >= K)
        sk.remove(A[k++])
      res += k - j
    }
    return res
  }

  @Test
  fun test1(){
    assertEquals(7, subarraysWithKDistinct(intArrayOf(1,2,1,2,3), 2))
  }

  @Test
  fun test2(){
    assertEquals(3, subarraysWithKDistinct(intArrayOf(1,2,1,3,4), 3))
  }
}